### PR TITLE
Add functionality for operators to template all eligible fields (apac…

### LINKED ISSUE #27285: Using .output on non-templated fields
### Discussed in https://github.com/apache/airflow/discussions/26938

<div type='discussions-op-text'>

<sup>Originally posted by **stephenonethree** October  7, 2022</sup>
I just discovered the `.output` property functionality that apparently was released in Airflow 2 for classic operators, as a simple way of accessing their output XComs. I think that this is a super useful feature because it would allow simpler connections between tasks than what I have been doing until now. 

Until now, I've been explicitly giving a downstream task the task_ids and XCom names that it needs to pull from its upstreams (as hardcoded string parameters). Something like:

```
# pushes XCom named myvar
upstream_task=SomeOperator(task_id='upstream_task')

# this is a custom operator and within execute() I have a line roughly like:
# actual_input = context['task_instance'].xcom_pull(
#     dag_id=context['dag'].dag_id, task_ids=upstream_task_id, key=upstream_xcom_name)
this_task = OtherOperator(
    task_id='this_task',
    upstream_task_id='upstream_task',
    upstream_xcom_name='myvar'
)
upstream_task >> this_task
```

With `.output` I could simplify this to:

```
# pushes XCom named myvar
upstream_task=SomeOperator(task_id='upstream_task')

this_task = OtherOperator(
    task_id='this_task',
    actual_input=upstream_task.output['myvar']
)
upstream_task >> this_task
```

Unfortunately it seems that there is one limitation. On the TaskFlow documentation page (https://airflow.apache.org/docs/apache-airflow/2.4.1/tutorial/taskflow.html#consuming-xcoms-between-decorated-and-traditional-tasks) it says: _Using the .output property as an input to another task is supported only for operator parameters listed as a template_field._

I don't use Jinja templating for very many of my parameters as it's mostly irrelevant for me. So my questions are, is there a technical reason for this limitation? If not, is this limitation something that you are considering dropping in a future Airflow version? I suppose I could just make these fields templated to get around it, but I don't really want to turn on templating if I don't expect to need it, since I suppose it introduces the possibility of incorrect interpolation (though perhaps that's a remote possibility because I don't think most of my variables will include `{{` or `}}`.

Let me know if I should file this as a feature request instead, for now I guess the Ideas category works.</div>

### PR DESCRIPTION
**Pull request by rkarish**
Closes: apache#27285
Related: apache#26938

### CODE REVIEW

**Review by potiuk — APPROVED** [review]
Nice. I think this is the right approach for the problem we've been discussing a number of times.

**Comment by potiuk** [comment]
cc: @uranusjr  - any particular concerns for that one (especially for mapped operators ?). I think we might need some specific exclusions ?

**Comment by potiuk** [comment]
(And tests and static checks need to be fixed BTW).

**Comment by rkarish** [comment]
Thanks Jarek, and please let me know if there is anything else that I can do for this, like additional documentation or example code. Also, per the failing static checks and tests, is this related to my changes and something that I need to personally resolve?

**Comment by potiuk** [comment]
But if you are completely lost after rebase and checking other PRs and cannot figure how your change caused it, you can always ask for help. Sometimes there are weird, indirect chain reactions that might cause it.

**Comment by rkarish** [comment]
Sounds good, I will investigate tonight when I have time and reach out in Slack if I need any assistance. Thanks Jarek.

**Comment by rkarish** [comment]
Hi Jarek, quick update on this. The failing static checks was an issue in main, it was resolved with a rebase. All other failures were related to my code changes and I've resolved those failing tests by updating some code in the serialization package.

File: airflow/models/baseoperator.py [hunk 1]
```diff
@@ -1184,7 +1197,29 @@ def render_template_fields(
         """
         if not jinja_env:
             jinja_env = self.get_template_env()
-        self._do_render_template_fields(self, self.template_fields, context, jinja_env, set())
+
+        if self.template_all_fields:
+            templatable_fields = (
```
Comment (rkarish) [inline airflow/models/baseoperator.py:1202]: Was wondering if anyone has any comment on how I did this. Goal was to get attributes that belong only to child classes of `BaseOperator`, would like to know if this approach is satisfactory, or if there is a better way to accomplish it.

File: airflow/models/baseoperator.py [hunk 2]
```diff
@@ -1184,7 +1197,29 @@ def render_template_fields(
...
+                set(self.__dict__.keys())
+                - set(BaseOperator.__dict__.keys())
+                - set(partial(operator_class=self.__class__, task_id=self.task_id).kwargs.keys())
+            )
```
Comment (rkarish) [inline airflow/models/baseoperator.py]: Was wondering if anyone has any comment on how I did this. Goal was to get attributes that belong only to child classes of `BaseOperator`, would like to know if this approach is satisfactory, or if there is a better way to accomplish it.

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
LGTM. @uranusjr - I think you were earlier involved in discussions about that one - WDYT?

**Comment by Taragolis** [comment]
Should we also do something with [`[core] max_num_rendered_ti_fields_per_task`](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#max-num-rendered-ti-fields-per-task) ?

File: airflow/models/baseoperator.py (hunk 2 above)
Comment (Taragolis) [reply airflow/models/baseoperator.py]: Well I guess better create some restricted frozen set and keep all parameters which never rendered (I think all of BaseOperator) we also could use this set as reference for end users. It is quite common question "Why I can't render `pool`, `executor_config`, `retries` and etc"

Comment (Taragolis) [reply airflow/models/baseoperator.py]: And I think we need to exclude all implicit attributes which started with `_`.

Comment (Taragolis) [reply airflow/models/baseoperator.py]: And idea for another PR: Deprecated all explicit `templated_fields` which started with `_`.

🙄 
https://github.com/apache/airflow/blob/98221da0d96b102b009d422870faf7c5d3d931f4/airflow/providers/apache/spark/operators/spark_submit.py#L75-L90

Comment (Taragolis) [reply airflow/models/baseoperator.py]: And just wondering is it work with properties?

```python
class OperatorWithProperties(BaseOperator):

    @property
    def hook(self):
        return "foo"

    @cached_property
    def cached_hook(self):
        return "bar"

    def execute(self, context):
        pass
```

I think we should skip all properties

**Comment by rkarish** [comment]
> Should we also do something with [`[core] max_num_rendered_ti_fields_per_task`](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#max-num-rendered-ti-fields-per-task) ?

Thanks for pointing this out, I will look into this configuration to see if there are any implications with the new functionality in these changes.

**Comment by rkarish** [comment]
@Taragolis for the `max_num_rendered_ti_fields_per_task` configuration I am proposing to make an update to the docstring for `BaseOperator` making users aware of this limitation, and if they enable this new feature they are limited by said configuration and may need to increase it. WDYT?

**Comment by Taragolis** [comment]
> @Taragolis for the max_num_rendered_ti_fields_per_task configuration I am proposing to make an update to the docstring for BaseOperator making users aware of this limitation, and if they enable this new feature they are limited by said configuration and may need to increase it. WDYT?


Oh.. I'm think I'm stupid 🤦 I understood description of configuration incorrectly. 
It is not affect any amount of templated fields during TaskIntance run, rather than total number rendered fields of particular task

**Comment by rkarish** [comment]
This one is good to go unless anyone has any objections 😄

File: airflow/models/baseoperator.py [hunk 3]
```diff
@@ -1184,7 +1199,23 @@ def render_template_fields(
...
         if not jinja_env:
             jinja_env = self.get_template_env()
-        self._do_render_template_fields(self, self.template_fields, context, jinja_env, set())
+
+        if self.template_all_fields:
+            template_fields = (
+                {key for key in self.__dict__.keys() if isinstance(getattr(self, key), str)}
+                - set(BaseOperator.__dict__.keys())
+                - set(MappedOperator.__dict__.keys())
+            )
```
Comment (uranusjr) [inline airflow/models/baseoperator.py:1208]: Does not looks right MappedOperator is relevant here, it breaks the OOP principles.

**Comment by uranusjr** [comment]
I have doubts making this an operator argument is the right design. I’d want to make this designed at the operator class level, e.g.

```python
class MyOperator(BaseOperator):
    template_fields = "__all__"  # Or we could introduce a flag constant for this.
    template_fields_excludes = [...]  # If needed.
```

And then we can work to convert existing operators gradually with more confidence they work.

File: airflow/models/baseoperator.py (hunk 3 above)
Comment (rkarish) [reply airflow/models/baseoperator.py:1208]: For this if I only included the attributes from `BaseOperator` then some attributes get leaked through from the `MappedOperator`. This is the result of test when not removing additional attributes from `MappedOperator`.

```{'weight_rule', 'arg3', 'queue', 'trigger_rule', 'owner', 'arg1', 'arg2', 'task_id'}```

With both all the attributes belonging to the parent classes are excluded.

```{'arg1', 'arg2', 'arg3'}```

**Comment by rkarish** [comment]
> I have doubts making this an operator argument is the right design. I’d want to make this designed at the operator class level, e.g.
> 
> ```python
> class MyOperator(BaseOperator):
>     template_fields = "__all__"  # Or we could introduce a flag constant for this.
>     template_fields_excludes = [...]  # If needed.
> ```
> 
> And then we can work to convert existing operators gradually with more confidence they work.

With this approach how are the fields to template determined?

**Comment by uranusjr** [comment]
Same way as it’s determined now (if `template_fields` is `__all__` on the class we calculate it in `__init__`).

**Comment by rkarish** [comment]
@potiuk WDYT? Should we just close this one out?

**Review by Taragolis — CHANGES_REQUESTED** [review]
I think we need more wide discussion about final implementation. @potiuk @uranusjr WDYT?

Note: Request changes only for protection of incidental merge, so potentially we could merge this implementation finally.

**Comment by potiuk** [comment]
> I have doubts making this an operator argument is the right design. I’d want to make this designed at the operator class level, e.g.

Agree with @uranusjr that class level field is better. Also we should allow (and document) the use of it in this form in DAG:


```python
BashOperator.template_all_fields = True
BashOperator.exclude_fields_from_template = []

with DAG(....
````

Then we would not have to convert all operators, but the user, when writing a DAG will not have to create a custom operator to template all fields.

**Pull request closed without merging**